### PR TITLE
[WIP] Expose map camera and allow direct transformations of modified annotations and annotation views

### PR DESCRIPTION
**Pull request by alex-reilly-pronto**
Hi! Thanks for maintaining this library. It's been a pleasure to use. I've got some annotations where I want to maintain their orientation relative to the map and move them over time, so I'm exposing the camera along with a closure that gives the user direct access to modified annotations and annotation views. 

The suggested method online for changing annotation location is to use a `MKPointAnnotation` rather than an `MKAnnotation`, and to alter the `coordinate` field and for changing annotation orientation is to do a `CGAffineTransform` on the view itself. I'm doing both, but I'm open to something else if you've got a suggestion. It does feel like this requires developers to touch UIKit more than you'd probably like.

Here's a snippet of it in action.
```swift
// ...
.modifiedAnnotations { modifiedAnnotation, mkannotation, view in
  mkannotation.coordinate = modification.coordinate
  view.transform = CGAffineTransform(
    rotationAngle: modifiedAnnotation.annotationHeading.degreesToRadians
  )
}
.camera(viewStore.binding(get: \.camera, send: MapAction.moveCamera))
```

Here we've got the camera bound to a field in my view's state, which is being used to calculate `annotationHeading` behind the scenes. 

Prior to this, if an annotation was modified it wouldn't be updated because `Map.Coordinator` was only checking for inserted or removed `id`s. Because of that, annotations are now required to be `Equatable` so that the coordinator can check for modified annotations in addition to inserted and removed annotations.

If we naively remove and then insert modified annotations, then that causes the annotation to flicker. 

### Breaking Changes
- annotation items are now required to be `Equatable`

### CODE REVIEW

File: Sources/Map/Map+Coordinator.swift [hunk 1]
```diff
@@ -94,15 +143,26 @@ extension Map {
                     }
                     annotationContentByID[item.id] = content
                     annotationContentByObject[objectKey] = content
+                    
                     registerAnnotationViewIfNeeded(on: mapView, for: content)
+              
                     mapView.addAnnotation(content.annotation)
+                    
+                    if let annotationView = content.view(for: mapView) {
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:151]: Figure out when this will be nil

File: Sources/Map/Map+Coordinator.swift [hunk 2]
```diff
@@ -94,15 +143,26 @@ extension Map {
                     }
                     annotationContentByID[item.id] = content
                     annotationContentByObject[objectKey] = content
+                    
                     registerAnnotationViewIfNeeded(on: mapView, for: content)
+
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:148]: remove extra lines

File: Sources/Map/Map+Coordinator.swift [hunk 3]
```diff
@@ -74,12 +75,60 @@ extension Map {
...
             } else {
-                changes = newView.annotationItems.difference(from: []) { $0.id == $1.id }
+                changes = newView.annotationItems.difference(from: []) { $0 == $1 }
             }
-
+            
+            var changeSet = Set<AnnotationItems.Element.ID>()
+            var modificationsSet = [AnnotationItems.Element]()
+            
+            var changesWithoutModifications = changes.map { $0 }
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:86]: Make sure changes isn't used past this point

File: Sources/Map/Map+Coordinator.swift [hunk 4]
```diff
@@ -94,15 +143,26 @@ extension Map {
...
+                    }
+                    
                 case let .remove(_, item, _):
                     guard let content = annotationContentByID[item.id] else {
                         assertionFailure("Missing annotation content for item \(item) encountered.")
                         continue
                     }
+                    
                     mapView.removeAnnotation(content.annotation)
+
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:163]: remove

File: Sources/Map/Map+Coordinator.swift [hunk 5]
```diff
@@ -248,6 +308,7 @@ extension Map {
             DispatchQueue.main.async { [weak self] in
                 self?.view?.coordinateRegion = mapView.region
                 self?.view?.mapRect = mapView.visibleMapRect
+                self?.view?.camera?.wrappedValue = mapView.camera
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:311]: make camera work as a two way binding

File: Sources/Map/Map+Coordinator.swift [hunk 6]
```diff
@@ -289,20 +350,15 @@ extension Map {
         }
 
         public func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
-            guard let content = annotationContentByObject[ObjectIdentifier(annotation)] else {
-                return nil
-            }
-            return content.view(for: mapView)
+            return viewByObject[ObjectIdentifier(annotation)]
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:353]: Document this

File: Sources/Map/Map.swift [hunk 7]
```diff
@@ -11,12 +11,13 @@ import MapKit
 import SwiftUI
 
 public struct Map<AnnotationItems: RandomAccessCollection, OverlayItems: RandomAccessCollection>
-    where AnnotationItems.Element: Identifiable, OverlayItems.Element: Identifiable {
+where AnnotationItems.Element: Identifiable, OverlayItems.Element: Identifiable, AnnotationItems.Element: Equatable {
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map.swift:14]: indent

File: Sources/Map/Map+Coordinator.swift [hunk 8]
```diff
@@ -74,12 +75,60 @@ extension Map {
...
+            
+            for modification in modificationsSet {
+                guard
+                    let content = annotationContentByID[modification.id],
+                    let annotationView = viewByObject[ObjectIdentifier(content.annotation)]
+                else {
+                    fatalError()
+                }
+                
+                view?.modifiedAnnotationHandler?(modification, content.annotation, annotationView)
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:128]: only do all this work if the handler exists

File: Sources/Map/Map.swift [hunk 9]
```diff
@@ -36,7 +37,20 @@ public struct Map<AnnotationItems: RandomAccessCollection, OverlayItems: RandomA
 
     let overlayItems: OverlayItems
     let overlayContent: (OverlayItems.Element) -> MapOverlay
-
+    
+    var modifiedAnnotationHandler: ((AnnotationItems.Element, MKPointAnnotation, MKAnnotationView) -> Void)?
+    
+    public func camera(_ binding: Binding<MKMapCamera>) -> Self {
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map.swift:43]: Document

File: Sources/Map/Map.swift [hunk 10]
```diff
@@ -36,7 +37,20 @@ public struct Map<AnnotationItems: RandomAccessCollection, OverlayItems: RandomA
...
+        var map = self
+        map.camera = binding
+        return map
+    }
+    
+    public func modifiedAnnotations(handler: @escaping (AnnotationItems.Element, MKPointAnnotation, MKAnnotationView) -> Void) -> Self {
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map.swift:49]: document

File: Sources/Map/Map+Coordinator.swift [hunk 11]
```diff
@@ -74,12 +76,42 @@ extension Map {
...
+                case let .insert(_, element, _):
+                    return element
+                case let .remove(_, element, _):
+                    return element
+                }
+            }
+            
+            let changeId = { change in
+                return changeElement(change).id
+            }
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:95]: Move these to private functions

File: Sources/Map/Map+Coordinator.swift [hunk 12]
```diff
@@ -74,12 +76,42 @@ extension Map {
...
+            
+            for modification in modificationsList {
+                if
+                    let content = annotationContentByID[modification.id],
+                    let annotationView = viewByObject[ObjectIdentifier(content.annotation)],
+                    let annotation = content.annotation as? MKPointAnnotation
+                {
+                    view?.modifiedAnnotationHandler?(modification, annotation, annotationView)
+                }
+            }
```
Comment (alex-reilly-pronto) [inline Sources/Map/Map+Coordinator.swift:111]: find a way to avoid doing all this if `modifiedAnnotationHandler` doesn't exist

Comment (pauljohanneskraft) [reply Sources/Map/Map+Coordinator.swift:111]: I might be able to help here - similarly to https://github.com/pauljohanneskraft/Map/issues/15 you might be able to create a publisher to change these properties.

You can use views like the following to create these publishers:

```swift
struct ChangeListener<Value: Equatable, Content: View>: View {

    // MARK: Stored Properties

    let value: Value
    @ViewBuilder var content: (PassthroughSubject<Value, Never>) -> Content

    @State private var subject = PassthroughSubject<Value, Never>()

    // MARK: Computed Properties

    var body: some View {
        content(subject)
            .onChange(of: value, perform: subject.send)
    }

    // MARK: Initialization

    init(for value: Value, @ViewBuilder content: @escaping (PassthroughSubject<Value, Never>) -> Content) {
        self.value = value
        self.content = content
    }

}
```

In many cases with these highly fluctuating and often-changing values, you will probably already have publishers (or async sequences) anyways in these cases, right? 😅

Comment (pauljohanneskraft) [reply Sources/Map/Map+Coordinator.swift:111]: Alternatively, you can also look into `ObservableObject` and how you might be able to use its `objectWillChangePublisher` that is triggered for each `@Published` property.

Comment (alex-reilly-pronto) [reply Sources/Map/Map+Coordinator.swift:111]: Thanks for your feedback! I'm a little confused though. How do you see `ChangeListener` being used?

Comment (alex-reilly-pronto) [reply Sources/Map/Map+Coordinator.swift:111]: I _think_ I get what you're going for actually. Create `ChangeListener`, use it to wrap the view being passed to a `ViewMapAnnotation`, and pass it a publisher that will be updated with a changing value so the wrapped view can be updated based on those changes.

Is that it? I've got to move on for now at work, but I'll come back to this PR at some point. Feel free to steal what's here if you want to, or close the PR in the meantime and I'll reopen it once I have time.

**Comment by alex-reilly-pronto** [comment]
Closing this out for now.

**Pull request closed without merging**